### ISSUE TITLE
redis streams 

### ISSUE THREAD

**Issue by fas3r** [fastify/fastify-redis#31]
Hello, 

someone in the ioredis community posted the necessary to add the redis streams command like xread, xreadgroup, xadd, xack, etc ... [in this post](https://github.com/luin/ioredis/issues/747#issuecomment-447008824). 

Using them for a while it's working fine and I actually use your plugin with the "client" option to  use this redis instance instead of the default one.

Do you think you could add those ? 

Thanks by advance.

**Comment by xoryouyou** [comment]
Would love to see Streams added since I am too using a home-cooked version right now.

**Comment by Eomm** [comment]
I can give it a check! Of course, if you have already clear in mind what to do, don't hesitate to send a PR 💪

**Comment by xoryouyou** [comment]
I am just using a modified version of this https://github.com/luin/ioredis/issues/747#issuecomment-445058656

Nothing PR worthy unfortunately.

**Comment by darkgl0w** [comment]
Hi,

I think it's a bit out of the scope of this package as this is just a graceful wrapper built on top of `ioredis` package and you should open an issue to ask for support of this feature or send a PR directly on `ioredis` repo.

But if @Eomm wants to add this functionality as an utility in `fastify-redis` I will be glad to offer my help on my free time.

Currently `ioredis` already exposes all the tools needed to build something like this (through the `sendCommand` and `Command` methods)

A little POC snippet here to show how stream related commands can be used through `ioredis` tools (it's a bit dirty, but that's just a demo so I hope you don't mind ^^') :
```js
// ioredis_stream_poc.js
'use strict';

const Redis = require('ioredis');
const redis = new Redis();

async function littleProofOfConcept () {
  // We write an event to the stream 'my awesome fastify stream name', setting 'key' to 'value'
  await redis.sendCommand(
    new Redis.Command('XADD', ['my awesome fastify stream name', '*', 'hello', 'fastify is awesome'])
  );

  // We read events from the beginning of the stream called 'my awesome fastify stream name'
  let response = await redis.sendCommand(
    new Redis.Command('XREAD', ['STREAMS', 'my awesome fastify stream name', 0])
  );

  // We parse the results
  let events = response[0][1];

  for (let i = 0; i < events.length; i++) {
    const e = events[i];
    console.log('#LOG: id is', e[0].toString());

    // We log each key
    for (const key in e[1]) {
      console.log(e[1][key].toString());
    }
  }

  redis.disconnect();
}

littleProofOfConcept();
```

if you execute this little snippet you should see something like this on your console :
```sh
$ node ioredis_stream_poc.js 
#LOG: id is 1557146102229-0
hello
fastify is awesome
```

Here I just took the [Redis relevant documentation](https://redis.io/commands/xadd) and then I executed raw commands through the tools provided by `ioredis`.

In other words we should be able to pack some utils in this package without much hassle (as a side note, for this demonstration I didn't provide support for Node 6) ^^

Feel free to ping me if you still want to work on this and need some help @Eomm ^^.

**Comment by fas3r** [comment]
hello @darkgl0w , 

But there are nothing to do, as indicated in my previous post, this [link](https://github.com/luin/ioredis/issues/747#issuecomment-447008824) already have all the necessary. Just the code need to be adapted as the constant 'ioredis' is name differently than the one in this module, that's all. 

Fas3r.

**Comment by Eomm** [comment]
I would like to specify that I didn't check the code yet, so I don't evaluate if add this to the plugin or not.

Looking the links, right now, @darkgl0w seems to have right, but for sure we can PR `ioredis` or build a new plugin out of the fastify org if it is not worth adding it.

for sure, @fas3r , it is not only a copy paste, because we must:
+ write docs
+ maintain the code through years
+ add test
+ fix bugs
+ reply to issues

I'm happy to help the community of fastify, for this reason, I said that I (will) give it a check 😃

**Comment by fas3r** [comment]
Hello @Eomm , 

Of course test, comment and doc need to be done, but @darkgl0w was describing how to do it, while the code is already done the way it's suppose to be in the suggested link, if you like to reinvent the wheel it's up to you :).

fas3r.

**Comment by mcollina** [comment]
I think the path forward is either:

- add support for redis streams to ioredis.
- create another module that adds that support, and make it easy to use it with fastify-redis.

I do not think that supporting this feature is in the scope of this repo.

**Comment by fas3r** [comment]
Hello @mcollina , 

They will not add to ioredis  label `wontfix` on the post that I shared.

**Comment by fas3r** [comment]
@darkgl0w  https://github.com/luin/ioredis/issues/747, should be release by june 20th.

**Comment by fas3r** [comment]
Hello @darkgl0w , @Eomm , it was release in [4.10.0](https://github.com/luin/ioredis/issues/875), could you update the dependency of ioredis ? 

Thanks

**Comment by darkgl0w** [comment]
Hello @fas3r ^^
I have a PR ready for this. I am just waiting a proper release of `ioredis` (currently the `4.10.0` version is tagged as `next` and `4.9.5` is considered the `latest` stable main release, as soon as it lands as `latest` stable I will submit the PR).

**Comment by fas3r** [comment]
Hello @darkgl0w , 

I asked and it set to the latest according to [this comment from one of the author](https://github.com/luin/ioredis/issues/875#issuecomment-500100747)

Fas3r.

**Comment by Eomm** [comment]
Good

I would like only to add that is already possible use that version:

```
mkdir npm-rulez
cd npm-rulez
npm init --yes
npm i fastify-redis
npm ls
# print out
`-- fastify-redis@3.1.0
  +-- fastify-plugin@1.6.0
  | `-- semver@6.1.1
  `-- ioredis@4.10.0
    +-- cluster-key-slot@1.0.12
    +-- debug@3.2.6
    | `-- ms@2.1.2
    +-- denque@1.4.1
    +-- lodash.defaults@4.2.0
    +-- lodash.flatten@4.4.0
    +-- redis-commands@1.5.0
    +-- redis-errors@1.2.0
    +-- redis-parser@3.0.0
    | `-- redis-errors@1.2.0 deduped
    `-- standard-as-callback@2.0.1
```

since that version is in range.

For a project with a package-lock should be possible to use `npm update --save`
